build-logic/src/main/groovy/qsl.module.gradle

```diff
@@ -0,0 +1,171 @@
+/*
+ * Buildscript for defining a QSL module.
+ *
+ * This is a convention plugin, and is applied by adding the following to your `plugins` block: `id("qsl.module")`.
+ *
+ * A module is a mod. More specifically the mod exposes a public API for mod developers to use.
+ *
+ * This buildscript declares an extension available under the `qslModule` name. This extension is used to configure
+ * attributes of the module, namely the name of the module, version, parent library, dependent libraries and any
+ * additional things to add to the quilt.mod.json. The buildscript also declares some additional tasks that can be used
+ * to run the game and runTestmodClient, runTestmodServer. These testmod tasks can be run using
+ * `gradle LIBRARY_NAME:MODULE_NAME:TASK_NAME`.
+ */
+
+import qsl.internal.Git
+import qsl.internal.extension.QSLModuleExtension
+
+plugins {
+	id("java-library")
+	id("maven-publish")
+	id("net.kyori.indra.git")
+	id("org.quiltmc.loom")
+	id("org.quiltmc.gradle-conventions.java.style")
+	id("org.quiltmc.gradle-conventions.java.version")
+	id("qsl.common")
+}
+
+def extension = project.extensions.create("qslModule", QSLModuleExtension, project)
+
+// We need to wait until project evaluation has finished to peek into a complete QSL module extension.
+afterEvaluate {
+	// Module name should also be the name of artifacts.
+	archivesBaseName = extension.moduleName
+
+	if (extension.library.get() == null) {
+		throw new GradleException("Module $extension.moduleName needs the `library` field set in the qslModule extension")
+	}
+
+	// Version is set automatically in extension, verify it is set.
+	if (project.version == "unspecified") { // Gradle will return `unspecified` given no version being set.
+		throw new GradleException("Module $extension.moduleName needs a version to be set in the qslModule extension")
+	}
+
+	// Apply module dependencies for maven.
+	extension.setupModuleDependencies()
+	// TODO: Anything else to validate.
+
+	group = "org.quiltmc.qsl-prototype.${extension.library.get()}"
+}
+
+sourceSets {
+	testmod {
+		compileClasspath += sourceSets.main.compileClasspath
+		runtimeClasspath += sourceSets.main.runtimeClasspath
+	}
+}
+
+dependencies {
+	// testmod sourceSet should depend on everything in the main source set.
+	testmodImplementation sourceSets.main.output
+}
+
+configurations {
+	// The reason for this `dev` configuration has been lost to time.
+	// If you happen to figure out why extension is needed, or a way to eliminate it, please tell us.
+	dev
+}
```

Comment:
Where is the configuration defined?

Reply:
It is defined right there

Reply:
Isn't this just telling Gradle to use it, though?

Reply:
The configurations block declare and defines the different configurations.

See https://docs.gradle.org/current/dsl/org.gradle.api.artifacts.ConfigurationContainer.html.